internal/controller/postgrescluster/pod_disruption_budget.go

```diff
@@ -0,0 +1,99 @@
+/*
+ Copyright 2021 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package postgrescluster
+
+// Note: The behavior for an empty selector differs between the
+// policy/v1beta1 and policy/v1 APIs for PodDisruptionBudgets. For
+// policy/v1beta1 an empty selector matches zero pods, while for
+// policy/v1 an empty selector matches every pod in the namespace.
+// https://kubernetes.io/docs/tasks/run-application/configure-pdb/#specifying-a-poddisruptionbudget
+import (
+	"github.com/pkg/errors"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
+)
+
+// generatePodDisruptionBudget takes parameters required to fill out a PDB and
+// returns the PDB
+func (r *Reconciler) generatePodDisruptionBudget(
+	cluster *v1beta1.PostgresCluster,
+	meta metav1.ObjectMeta,
+	minAvailable int32,
+	selector metav1.LabelSelector,
+) (*policyv1beta1.PodDisruptionBudget, error) {
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: meta,
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			// We can't pass the IntOrString value directly from our spec
+			// into the PDB spec because of restrictions placed on Arbitrary
+			// Controllers.
+			// - only .spec.minAvailable can be used, not .spec.maxUnavailable.
+			// - only an integer value can be used with .spec.minAvailable, not
+			// a percentage.
+			// https://kubernetes.io/docs/tasks/run-application/configure-pdb/#arbitrary-controllers-and-selectors
```

Comment:
:thinking: The "Arbitrary Controllers and Selectors" section in the docs is throwing me off a bit.  Specifically, I'm not sure if this applies to us considering our controller does not actually _own_ the Pods, but rather PGO owns the StatefulSet or Deployment that owns the Pods (or the resource that then creates the Pods, i.e. a ReplicaSet in the case of a Deployment).

And looking at [other parts of the PDB docs](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) I see that it actually discovers the owning workload for the Pod, which wouldn't be us (but rather a StatefulSet or ReplicaSet):

> The control plane discovers the owning workload resource by examining the .metadata.ownerReferences of the Pod.

Additionally, I did an experiment (this was with a local Kube 1.22 cluster) where I tried violating both of the restrictions described for the PDB.  Specifically, I tried setting `minAvailable` as a percentage and also tried setting `maxUnavialable` (i.e. instead of `minAvailable`), and the PDB appeared to work.

For instance, here is the PDB with `maxUnavailable`:

```bash
$ kubectl get pdb
NAME        MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
instance1   N/A             1                 0                     126m
```

And when I go to drain I only see one of my two PG replicas terminate, and the PDB is respected:

```bash
error when evicting pods/"hippo-instance1-pk26-0" -n "postgres-operator" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

Maybe the limitations don't apply to us since we do not directly own the Pods?  Though if anything I am missing here, please let me know.

Reply:
@andrewlecuyer I think [this is the code](https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/controller/disruption/disruption.go#L665) that finds the specified/expected replicas/count, and you're right it is finding each controller for each pod and doing a sum.

This is the [line of the docs](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#arbitrary-controllers-and-selectors) that had me concerned (emphasis mine):
> You can skip this section if you only use PDBs with the built-in application controllers (Deployment, ReplicationController, ReplicaSet, and StatefulSet), with **the PDB selector matching the controller's selector**.

Reply:
Yup, completely agree - that line gave me pause too!

And thanks for pulling that code, great to verify what they're doing there down at that level.

And after seeing that code, I get the feeling the doc isn't 100% accurate ... i.e. thinking it really should say something like:

> ...with the PDB selector selecting Pods that are also selected by one or more built-in application controllers

(or something like that ... feels like to much use of "select" in there :laughing:)

Reply:
@andrewlecuyer Were you using the betav1 version of policy/PDBs? I wonder if the behavior would be different. They note that there is a difference in how selectors work (specifically the empty selector).

I'm not fully tracking what change is necessary here. Should I update the comment or are we talking about a code change?

Reply:
Updated to pass the `IntOrString` value through from the spec to the PDB